# -*- coding: utf-8 -*-

# DO NOT EDIT THIS FILE!
# This file has been autogenerated by dephell <3
# https://github.com/dephell/dephell

try:
    from setuptools import setup
except ImportError:
    from distutils.core import setup

readme = ''

setup(
    long_description=readme,
    name='biopipen',
    version='0.0.1',
    description='A set of PyPPL processes for bioinformatics.',
    python_requires='==3.*,>=3.6.0',
    author='pwwang',
    author_email='pwwang@pwwang.com',
    license='MIT',
    entry_points={
        "console_scripts": [
            "biopipen = biopipen.cli:main",
            "biopipen-stats = biopipen.cli.proxy:stats",
            "biopipen-plink = biopipen.cli.proxy:plink",
            "biopipen-imtherapy = biopipen.cli.proxy:imtherapy",
            "biopipen-tabix = biopipen.cli.proxy:tabix",
            "biopipen-gatk = biopipen.cli.proxy:gatk",
            "biopipen-gene = biopipen.cli.proxy:gene",
            "biopipen-bedtools = biopipen.cli.proxy:bedtools",
            "biopipen-resource = biopipen.cli.proxy:resource",
            "biopipen-gsea = biopipen.cli.proxy:gsea",
            "biopipen-rank = biopipen.cli.proxy:rank",
            "biopipen-hic = biopipen.cli.proxy:hic",
            "biopipen-power = biopipen.cli.proxy:power",
            "biopipen-tcgamaf = biopipen.cli.proxy:tcgamaf",
            "biopipen-snparray = biopipen.cli.proxy:snparray",
            "biopipen-rnaseq = biopipen.cli.proxy:rnaseq",
            "biopipen-genomeplot = biopipen.cli.proxy:genomeplot",
            "biopipen-mlearn = biopipen.cli.proxy:mlearn",
            "biopipen-fastx = biopipen.cli.proxy:fastx",
            "biopipen-picard = biopipen.cli.proxy:picard",
            "biopipen-eqtl = biopipen.cli.proxy:eqtl",
            "biopipen-common = biopipen.cli.proxy:common",
            "biopipen-cluster = biopipen.cli.proxy:cluster",
            "biopipen-plot = biopipen.cli.proxy:plot",
            "biopipen-tcga = biopipen.cli.proxy:tcga",
            "biopipen-sambam = biopipen.cli.proxy:sambam",
            "biopipen-vcf = biopipen.cli.proxy:vcf",
            "biopipen-cnvkit = biopipen.cli.proxy:cnvkit",
            "biopipen-tumhet = biopipen.cli.proxy:tumhet",
            "biopipen-sql = biopipen.cli.proxy:sql",
            "biopipen-bwtool = biopipen.cli.proxy:bwtool",
            "biopipen-bed = biopipen.cli.proxy:bed",
            "biopipen-misc = biopipen.cli.proxy:misc",
            "biopipen-seq = biopipen.cli.proxy:seq",
            "biopipen-network = biopipen.cli.proxy:network",
            "biopipen-tsv = biopipen.cli.proxy:tsv",
            "biopipen-snp = biopipen.cli.proxy:snp",
            "biopipen-tfbs = biopipen.cli.proxy:tfbs",
            "biopipen-marray = biopipen.cli.proxy:marray",
            "biopipen-web = biopipen.cli.proxy:web",
            "biopipen-chipseq = biopipen.cli.proxy:chipseq",
            "biopipen-algorithm = biopipen.cli.proxy:algorithm"
        ]
    },
    packages=[
        'biopipen', 'biopipen.cli', 'biopipen.cli.commands', 'biopipen.scripts',
        'biopipen.sets', 'biopipen.utils'
    ],
    package_dir={"": "."},
    package_data={
        "biopipen": [
            "*.toml", "reports/bcftools/*.md", "reports/gsea/*.md",
            "reports/imtherapy/*.md", "reports/mlearn/*.md",
            "reports/plot/*.md", "reports/rnaseq/*.md", "reports/stats/*.md",
            "reports/tcgamaf/*.md", "reports/tumhet/*.md"
        ],
        "biopipen.cli": ["*.toml"],
        "biopipen.cli.commands": ["*.toml"],
        "biopipen.scripts": [
            "algorithm/*.R", "algorithm/*.r", "cluster/*.r", "eqtl/*.R",
            "genomeplot/*.r", "gsea/*.R", "gsea/*.r", "marray/*.r", "misc/*.r",
            "mlearn/*.R", "mlearn/*.r", "plink/*.R", "plink/*.r", "plot/*.R",
            "plot/*.r", "power/*.r", "rank/*.r", "rnaseq/*.R", "rnaseq/*.r",
            "sambam/*.r", "stats/*.R", "stats/*.r", "tcga/*.r", "tcgamaf/*.R",
            "tcgamaf/*.bash", "tcgamaf/*.r", "tfbs/*.R", "tsv/*.R", "tsv/*.r",
            "tumhet/*.R"
        ],
        "biopipen.utils": ["*.R", "*.bash", "*.r", "*.toml"]
    },
    install_requires=[
        'diot', 'pyparam', 'pyppl==3.*', 'pyppl-annotate', 'pyppl-export',
        'pyppl-report', 'rich==6.*', 'toml==0.10.0'
    ],
)
